src/base/bittorrent/tracker.cpp

```diff
@@ -32,235 +33,410 @@
 
-    // 6. no_peer_id (extension)
-    announceReq.noPeerId = false;
-    if (queryParams.contains("no_peer_id"))
-        announceReq.noPeerId = true;
+    announceReq.peer.claimedPort = portNum;
 
-    // 7. TODO: support "compact" extension
+    // 4. numwant
+    const auto numWantIter = queryParams.find(ANNOUNCE_REQUEST_NUM_WANT);
+    if (numWantIter != queryParams.end()) {
+        const int num = numWantIter->toInt();
+        if (num < 0)
+            throw TrackerError("Invalid \"numwant\" parameter");
+        announceReq.numwant = num;
+    }
 
-    // Done parsing, now let's reply
-    if (announceReq.event == "stopped") {
+    // 5. no_peer_id
+    // non-formal extension
+    announceReq.noPeerId = (queryParams.value(ANNOUNCE_REQUEST_NO_PEER_ID) == "1");
+
+    // 6. left
+    announceReq.peer.isSeeder = (queryParams.value(ANNOUNCE_REQUEST_LEFT) == "0");
+
+    // 7. compact
+    announceReq.compact = (queryParams.value(ANNOUNCE_REQUEST_COMPACT) != "0");
+
+    // 8. cache `peers` field so we don't recompute when sending response
+    const QHostAddress claimedIPAddress {QString(announceReq.claimedAddress)};
+    announceReq.peer.endpoint = toBigEndianByteArray(!claimedIPAddress.isNull() ? claimedIPAddress : announceReq.address)
+        .append(static_cast<char>((announceReq.peer.claimedPort >> 8) & 0xFF))
+        .append(static_cast<char>(announceReq.peer.claimedPort & 0xFF))
+        .toStdString();
+
+    // 9. cache `ip` field so we don't recompute when sending response
+    announceReq.peer.ip = !announceReq.claimedAddress.isEmpty()
+        ? announceReq.claimedAddress.constData()
+        : announceReq.address.toString().toLatin1().constData(),
+
+    // 10. event
+    announceReq.event = queryParams.value(ANNOUNCE_REQUEST_EVENT);
+
+    if (announceReq.event.isEmpty()
+        || (announceReq.event == ANNOUNCE_REQUEST_EVENT_EMPTY)
+        || (announceReq.event == ANNOUNCE_REQUEST_EVENT_COMPLETED)
+        || (announceReq.event == ANNOUNCE_REQUEST_EVENT_STARTED)
+        || (announceReq.event == ANNOUNCE_REQUEST_EVENT_PAUSED)) {
+        // [BEP-21] Extension for partial seeds (partial support)
+        registerPeer(announceReq);
+        prepareAnnounceResponse(announceReq);
+    }
+    else if (announceReq.event == ANNOUNCE_REQUEST_EVENT_STOPPED) {
         unregisterPeer(announceReq);
     }
     else {
-        registerPeer(announceReq);
-        replyWithPeerList(announceReq);
+        throw TrackerError("Invalid \"event\" parameter");
     }
 }
 
 void Tracker::registerPeer(const TrackerAnnounceRequest &announceReq)
 {
-    if (announceReq.peer.port == 0) return;
-
     if (!m_torrents.contains(announceReq.infoHash)) {
-        // Unknown torrent
-        if (m_torrents.size() == MAX_TORRENTS) {
-            // Reached max size, remove a random torrent
+        // Reached max size, remove a random torrent
+        if (m_torrents.size() >= MAX_TORRENTS)
             m_torrents.erase(m_torrents.begin());
```

Comment:
The same comment as for peer deletion.
Maybe track last access time and delete based on it?